saleor/checkout/tests/test_utils.py

```diff
@@ -0,0 +1,52 @@
+from decimal import Decimal
+
+import pytest
+from prices import Money, TaxedMoney
+
+from ..utils import get_taxed_undiscounted_price
+
+BASE = Money("35.00", "USD")
+
+
+@pytest.mark.parametrize(
+    "price, tax_rate, prices_entered_with_tax, result",
+    [
+        # result should not be calculated but taken from price
+        (
+            TaxedMoney(gross=Money("43.74", "USD"), net=BASE),
+            Decimal(0.25),
+            False,
+            TaxedMoney(gross=Money("43.74", "USD"), net=BASE),
+        ),
+        # result should be calculated and different from price
+        (
+            TaxedMoney(gross=Money("43.74", "USD"), net=Money("36.00", "USD")),
+            Decimal(0.25),
+            False,
+            TaxedMoney(gross=Money("43.75", "USD"), net=BASE),
+        ),
+        # result should not be calculated and taken from price
+        (
+            TaxedMoney(gross=BASE, net=Money("26.26", "USD")),
+            Decimal(0.25),
+            True,
+            TaxedMoney(gross=BASE, net=Money("26.26", "USD")),
+        ),
+        # result should be calculated and different from price
+        (
+            TaxedMoney(gross=Money("36.00", "USD"), net=Money("26.26", "USD")),
+            Decimal(0.25),
+            True,
+            TaxedMoney(gross=BASE, net=Money("28.00", "USD")),
+        ),
+    ],
+)
+def test_get_taxed_undiscounted_price(price, tax_rate, prices_entered_with_tax, result):
+    result_price = get_taxed_undiscounted_price(
+        undiscounted_base_price=BASE,
+        price=price,
+        tax_rate=tax_rate,
+        prices_entered_with_tax=prices_entered_with_tax,
+    )
+
+    assert result_price == result
```

Comment:
IMO, instead of hardcoding the result, it should be calculated here based on the `tax_rate`. Otherwise the test is kind of magic and hard to maintenance.

Reply:
It kinda is, I agree. Though I had no other Idea how to test with `parameterize` all the scenarios for this function  - it checks and takes into consideration the corner-case when `tax_app` is returning slightly different results than local saleor calculations based on `tax_rate` ([this comment](https://github.com/saleor/saleor/pull/13585/files/8368e1944d6cd2b4ad8f741962864974b5203f00#diff-2f35ab82d7f669cedfbd9a04be1c39c1a1c3ee299730d7f0470254f5cc37721fR969)) 🤔 .

Reply:
Firstly this test probably should be split into two separates - one for checking when `prices_entered_with_tax` is `True`, when when it's `False`. Then as a prameters you can provide the `price` (I see that `tax_rate` is the same in all examples so there is no reason to parametrize it) and based on `tax_rate` we should calculate the `net` or `gross` amount and compare it with `result_price`